website_adv_image_optimization/controllers/main.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2018 Onestein
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import io
+import base64
+import json
+
+from odoo import http, tools
+from odoo.http import request
+from PIL import Image
+
+
+class MainController(http.Controller):
+    @http.route('/website_adv_image_optimization/attachment/optimize',
+                type='http', auth='user', methods=['POST'])
+    def optimize(self, attachment_id, quality=80,
+                 width=False, height=False, **kwargs):
+        attachment_obj = request.env['ir.attachment']
+        attachment = attachment_obj.search([('id', '=', attachment_id)])
+
+        image = Image.open(io.BytesIO(base64.b64decode(attachment.datas)))
+        opt = dict(format=image.format)
+        opt.update(optimize=True, quality=int(quality))
+        # Only resize when needed
+        width = int(width)
+        height = int(height)
+        if width and height and \
+           width > 0 and height > 0 and \
+           (width != image.size[0] or height != image.size[1]):
+            image.thumbnail((width, height), Image.ANTIALIAS)
+        img = io.BytesIO()
+        image.save(img, **opt)
+        data = img.getvalue()
+
+        # Not nice but needed to let GUI refresh
+        if '_adv_opt' not in attachment.name:
+            new_name = attachment.name + '_adv_opt(1)'
+        else:
+            count = str(int(attachment.name[-2:-1]) + 1)
+            new_name = attachment.name[0:-2] + count + ')'
+
+        attachment.write({
+            'datas': base64.b64encode(data),
+            'name': new_name
+        })
+        return json.dumps(True)
+
+    @http.route('/website_adv_image_optimization/attachment/dimensions',
+                type='http', auth='user', methods=['POST'])
+    def dimensions(self, attachment_id, **kwargs):
+        attachment_obj = request.env['ir.attachment']
+        attachment = attachment_obj.search([('id', '=', attachment_id)])
+
+        image = Image.open(io.BytesIO(base64.b64decode(attachment.datas)))
```

Comment:
As far as I see you use this only to get the image size to be used in `set_defaults`: can't you simply get the size via JS?

In any case you are leaving opened file objects around ;)

Reply:
I can try